File: apps/api/internal/interface/http/dto/sync.go
package dto

import (
	"time"

	"github.com/go-playground/validator/v10"
	"github.com/google/uuid"
	applicationdto "github.com/jeheskielSunloy77/libra-link/internal/application/dto"
	"github.com/jeheskielSunloy77/libra-link/internal/domain"
)

type StoreSyncEventRequest struct {
	EntityType      domain.SyncEntityType `json:"entityType" validate:"required,oneof=progress annotation bookmark preference reader_state"`
	EntityID        uuid.UUID             `json:"entityId" validate:"required"`
	Operation       domain.SyncOperation  `json:"operation" validate:"required,oneof=upsert delete"`
	Payload         map[string]any        `json:"payload"`
	BaseVersion     *int64                `json:"baseVersion" validate:"omitempty,gte=1"`
	ClientTimestamp *time.Time            `json:"clientTimestamp"`
	IdempotencyKey  string                `json:"idempotencyKey" validate:"required,min=8,max=255"`
}

func (d *StoreSyncEventRequest) Validate() error {
	return validator.New().Struct(d)
}

func (d *StoreSyncEventRequest) ToUsecase() *applicationdto.StoreSyncEventInput {
	clientTS := time.Now().UTC()
	if d.ClientTimestamp != nil {
		clientTS = d.ClientTimestamp.UTC()
	}

	return &applicationdto.StoreSyncEventInput{
		EntityType:      d.EntityType,
		EntityID:        d.EntityID,
		Operation:       d.Operation,
		Payload:         d.Payload,
		BaseVersion:     d.BaseVersion,
		ClientTimestamp: clientTS,
		IdempotencyKey:  d.IdempotencyKey,
	}
}

func parseSinceQuery(raw string) (*time.Time, error) {
	if raw == "" {
		return nil, nil
	}

	parsed, err := time.Parse(time.RFC3339, raw)
	if err != nil {
		return nil, err
	}
	utc := parsed.UTC()
	return &utc, nil
}

File: packages/zod/src/sync.ts
import { z } from 'zod'
import { ZGetManyQuery } from './utils.js'

export const ZSyncEntityType = z.enum([
	'progress',
	'annotation',
	'bookmark',
	'preference',
	'reader_state',
])
export const ZSyncOperation = z.enum(['upsert', 'delete'])

export const ZSyncEvent = z.object({
	id: z.string().uuid(),
	userId: z.string().uuid(),
	entityType: ZSyncEntityType,
	entityId: z.string().uuid(),
	operation: ZSyncOperation,
	payload: z.record(z.any()).optional(),
	baseVersion: z.number().int().optional(),
	clientTimestamp: z.string().datetime(),
	serverTimestamp: z.string().datetime(),
	idempotencyKey: z.string(),
	createdAt: z.string().datetime(),
})

export const ZStoreSyncEventDTO = z.object({
	entityType: ZSyncEntityType,
	entityId: z.string().uuid(),
	operation: ZSyncOperation,
	payload: z.record(z.any()).optional(),
	baseVersion: z.number().int().optional(),
	clientTimestamp: z.string().datetime().optional(),
	idempotencyKey: z.string().min(8),
})

export const ZSyncEventsQuery = ZGetManyQuery.extend({
	since: z.string().datetime().optional(),
})
